pkg/storage/s3/s3.go

```diff
@@ -76,38 +103,127 @@ func (d *driver) Volumes() ([]corev1.Volume, []corev1.VolumeMount, error) {
+	if d.Config.Bucket == "" {
+		return false, nil
+	}
+
+	svc, err := d.getSVC()
+	if err != nil {
+		return false, err
+	}
+	_, err = svc.HeadBucket(&s3.HeadBucketInput{
 		Bucket: aws.String(d.Config.Bucket),
 	})
-	return err
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchBucket:
+				util.UpdateCondition(cr, opapi.StorageExists, operatorapi.ConditionFalse, aerr.Code(), aerr.Error())
+				return false, nil
+			default:
+				return false, err
+			}
+		}
+		util.UpdateCondition(cr, opapi.StorageExists, operatorapi.ConditionUnknown, "Unknown Error Occurred", err.Error())
+		return false, err
+	}
+
+	util.UpdateCondition(cr, opapi.StorageExists, operatorapi.ConditionTrue, "S3 Bucket Exists", "")
+	return true, nil
+
 }
 
-// createBucket attempts to create an s3 bucket with the given name
-func (d *driver) createAndTagBucket(svc *s3.S3, installConfig *installer.InstallConfig, customResourceStatus *opapi.ImageRegistryStatus) error {
-	createBucketInput := &s3.CreateBucketInput{
-		Bucket: aws.String(d.Config.Bucket),
+// StorageChanged checks to see if the name of the storage medium
+// has changed
+func (d *driver) StorageChanged(cr *opapi.ImageRegistry) bool {
+	var storageChanged bool
+
+	if cr.Status.Storage.State.S3 == nil {
+		storageChanged = true
+	} else {
+		if cr.Status.Storage.State.S3.Bucket != cr.Spec.Storage.S3.Bucket {
+			storageChanged = true
+		}
 	}
 
-	// Create the S3 bucket
-	if _, err := svc.CreateBucket(createBucketInput); err != nil {
+	util.UpdateCondition(cr, opapi.StorageExists, operatorapi.ConditionUnknown, "S3 Bucket Changed", "S3 bucket is in an unknown state")
+
+	return storageChanged
+}
+
+func (d *driver) GetStorageName(cr *opapi.ImageRegistry) (string, error) {
+	if cr.Spec.Storage.S3 != nil {
+		return cr.Spec.Storage.S3.Bucket, nil
+	}
+	return "", fmt.Errorf("unable to retrieve bucket name from image registry resource: %#v", cr.Spec.Storage)
+}
+
+// CreateStorage attempts to create an s3 bucket with the given name
+// and apply any provided tags
+func (d *driver) CreateStorage(cr *opapi.ImageRegistry) error {
+	svc, err := d.getSVC()
+	if err != nil {
 		return err
 	}
 
+	ic, err := util.GetInstallConfig()
+	if err != nil {
+		return err
+	}
+
+	cv, err := util.GetClusterVersionConfig()
+	if err != nil {
+		return err
+	}
+
+	for {
+		if cr.Spec.Storage.S3.Bucket == "" {
+			d.Config.Bucket = fmt.Sprintf("%s-%s-%s-%s", clusterconfig.StoragePrefix, d.Config.Region, strings.Replace(string(cv.Spec.ClusterID), "-", "", -1), strings.Replace(string(uuid.NewUUID()), "-", "", -1))[0:62]
+		}
+
+		createBucketInput := &s3.CreateBucketInput{
+			Bucket: aws.String(d.Config.Bucket),
+		}
+
+		if _, err := svc.CreateBucket(createBucketInput); err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				switch aerr.Code() {
+				case s3.ErrCodeBucketAlreadyExists:
+					if cr.Spec.Storage.S3.Bucket != "" {
+						break
+					}
+					continue
```

Comment:
potential infinite loops scare me.  Put a cap of 5000 on this or something and update the status/condition to indicate we failed to generate a unique bucket name.  They can always restart the operator if they want another 5000 runs at it.

Reply:
updated in next push